chain/network/src/announce_accounts/mod.rs

```diff
@@ -0,0 +1,108 @@
+use crate::store;
+use lru::LruCache;
+use near_primitives::network::{AnnounceAccount, PeerId};
+use near_primitives::types::AccountId;
+use parking_lot::Mutex;
+use std::collections::HashMap;
+
+#[cfg(test)]
+mod tests;
+
+const ANNOUNCE_ACCOUNT_CACHE_SIZE: usize = 10_000;
+
+struct Inner {
+    /// Maps an account_id to a peer owning it.
+    account_peers: LruCache<AccountId, AnnounceAccount>,
+    /// Subset of account_peers, which we have broadcasted to the peers.
+    /// It is used to skip rebroadcasting the same data multiple times.
+    /// It contains less entries than account_peers in case some AnnounceAccounts
+    /// have been loaded from storage without broadcasting.
+    account_peers_broadcasted: LruCache<AccountId, AnnounceAccount>,
+    /// Access to store on disk
+    store: store::Store,
+}
+
+impl Inner {
+    /// Get AnnounceAccount for the given AccountId.
+    fn get_announce(&mut self, account_id: &AccountId) -> Option<AnnounceAccount> {
+        if let Some(aa) = self.account_peers.get(account_id) {
+            return Some(aa.clone());
+        }
+        match self.store.get_account_announcement(&account_id) {
+            Err(e) => {
+                tracing::warn!(target: "network", "Error loading announce account from store: {:?}", e);
+                None
+            }
+            Ok(None) => None,
+            Ok(Some(a)) => {
+                self.account_peers.put(account_id.clone(), a.clone());
+                Some(a)
+            }
+        }
+    }
+}
+
+pub(crate) struct Cache(Mutex<Inner>);
+
+impl Cache {
+    pub fn new(store: store::Store) -> Self {
+        Self(Mutex::new(Inner {
+            account_peers: LruCache::new(ANNOUNCE_ACCOUNT_CACHE_SIZE),
+            account_peers_broadcasted: LruCache::new(ANNOUNCE_ACCOUNT_CACHE_SIZE),
+            store,
+        }))
+    }
+
+    /// Adds accounts to the cache.
+    /// Returns the diff: new values that should be broadcasted.
+    /// Note: There is at most one peer id per account id.
+    pub(crate) fn add_accounts(&self, aas: Vec<AnnounceAccount>) -> Vec<AnnounceAccount> {
+        let mut inner = self.0.lock();
+        let mut res = vec![];
+        for aa in aas {
+            // We skip broadcasting stuff that is already broadcasted.
+            if inner.account_peers_broadcasted.get(&aa.account_id).map(|x| &x.epoch_id)
+                == Some(&aa.epoch_id)
+            {
+                continue;
+            }
+            inner.account_peers.put(aa.account_id.clone(), aa.clone());
+            inner.account_peers_broadcasted.put(aa.account_id.clone(), aa.clone());
```

Comment:
A tiny bit confusing that you are adding the accounts to broadcasted before actually broadcasting. Would that correctly handle any errors in the broadcasting logic? I suppose we'll retry broadcast next epoch and I have no clue if that is an issue or not.

Reply:
I don't really fully understand the purpose of this behavior, and I would propose to file a separate issue to investigate 
and possibly re-design how the announce accounts cache works (whereas this current PR just moves it without changing how it works).

Reply:
I don't entirely understand the purpose of this behavior. I would propose filing a separate issue to investigate 
and possibly re-design how the announce accounts cache works (whereas this current PR just moves it without changing how it works).